src/internal/watch/watch.go

```diff
@@ -141,6 +141,14 @@ func NewEtcdWatcher(ctx context.Context, client *etcd.Client, trimPrefix, prefix
 				// use new "nextRevision"
 				internalWatcher = etcd.NewWatcher(client)
 				rch = internalWatcher.Watch(ctx, prefix, watchOptions(nextRevision)...)
+				// We interpret a response with nil Error and 0 events as a server side cancellation
+				// as described in https://github.com/etcd-io/etcd/blob/client/v3.5.1/client/v3/watch.go#L53
+				//
+				// "If the context "ctx" is canceled or timed out, returned "WatchChan" is closed,
+				// and "WatchResponse" from this closed channel has zero events and nil "Err()"."
+				if resp.Err() == nil && len(resp.Events) == 0 {
```

Comment:
Isn't the `ok` variable above more direct evidence the channel has been closed?

Reply:
it is, but the way the updated etcd client signals that the ctx was cancelled on the server-side is by including `Err()==nil  && len(resp.Events) == 0`